rfcs/20231025-reusable-kernels/README.md

```diff
@@ -0,0 +1,131 @@
+
+Option 4 (primitive attribute) is the preferred option by frameworks, as they
+don't know which dimensions should be considered dynamic at runtime. It
+provides a balance between user control, flexibility, and ease of integration.
+Compared to the build flag it's more flexible, and compared to the environment
+variable it's less opaque.
+
+### Option 1: Using `DNNL_RUNTIME_DIM_VAL`
+
+With this option, the user will select one or more dimensions to have the
+`DNNL_RUNTIME_DIM_VAL` special value during primitive creation. oneDNN will
+then dispatch to the first implementation that supports the given set of
+runtime dimensions. One downside to this approach is that the implementation
+must either compile a kernel that's general enough to support any possible
+value for the runtime dimension (which limits optimization potential) or it
+must identify and compile all specialized kernels which support the given set
+of runtime dimensions (which would increase primitive creation time).
+
+In addition to existing limitations on primitives that support runtime
+dimensions, feedback from frameworks has been that which dimensions should be
+specified as `DNNL_RUNTIME_DIM_VAL` is unknown at primitive creation time.
+Between these two factors, this option is unlikely to lead to a satisfactory
+solution, at least in the short term.
+
+<u>Pros:</u>
+- The most fine-grained approach
+- Ties into existing mechanisms
+- Can be toggled without a rebuild
+
+<u>Cons:</u>
+- Several primitives have no implementations that support runtime dims -
+cannot be adopted until such implementations have been developed
+- Support for all possible runtime dim configurations is difficult
+- Requires users to know which dimensions will have dynamic shapes ahead of
+time
+
+### Option 2: Build flag
+
+With this option, the build flag `ONEDNN_PREFER_REUSABLE_KERNELS` would be
+added to select the preferred functionality. Default is `off`, and reusable
+kernels will be chosen preferentially when set to `on`.
+<u>Pros:</u>
+- Easy integration
+
+<u>Cons:</u>
+- Requires a rebuild to change the option
+- The setting is not apparent in the code, leads to opaque behaviour
+- Requires internal changes to detect implementation reusability
+
+### Option 3: Environment variable
+
+With this option, the user can set `ONEDNN_PREFER_REUSABLE_KERNELS=1` as an
+environment variable, which will cause the library to select reusable kernels
+preferentially. This option makes for easy integration, and provides a way to
+expand upon the functionality in the future (by setting a different value), for
+example if different levels of "reusability" are defined.
+<u>Pros:</u>
+- Easy integration
+- Allows for future expansion
+- Can be toggled without a rebuild
+
+<u>Cons:</u>
+- The setting is not apparent in the code, leads to opaque behaviour
+- Requires internal changes to detect implementation reusability
+
+### Option 4: Primitive attribute (preferred)
```

Comment:
Let's use this comment to discuss @mgouicem's summary comment. For the future, maybe it would better to leave a summary at a random place so that we can have a discussion by replying to it.

I second Mourad's point that the frameworks such as PyTorch and TensorFlow do not know which dimensions are dynamic and therefore they can either always enable the re-usable kernels or always disable them.
I think there are three use cases that we have seen so far:
1. Static dimensions - primitive/kernel cache helps to reduce creation overhead
2. Dynamic dimensions with a finite number of unique shapes - primitive/kernel cache with increased capacity helps to reduce creation overhead
3. Dynamic dimensions with undefined number of unique shapes - no solution

If the frameworks decide to always enable the re-usable kernels then they will see performance degradations for the first 2 scenarios. Is it something that have agreed on?


As for the @mgouicem's idea on asynchronous kernel creation, it's an interesting suggestion but it may create problems on its own. First of all in order for the idea to work we need a kernel cache that we already have, which is good. But with that, I think it may not be realistic to define the behavior for the users so that they know what to expect. For example, what if the specialized kernel gets evicted? Do we switch to the generic one or do we create the specialized kernel again? What if the generic kernel gets evicted? What if they both get evicted? How do we ensure deterministic behavior?

Reply:
> For example, what if the specialized kernel gets evicted? Do we switch to the generic one or do we create the specialized kernel again? 

The specialized kernel gets evicted only only if all primitive objects using it are destroyed. Which means for next primitive creation that would need it, the same flow applies:
- primitive creation triggers generic and specialized kernel jitting. Because all calls to convolution primitive creation trigger generic kernel compilation, it will very likely be resident in the kernel cache.
- during primitive execution, if the specialized kernel is ready, it gets used, if not it uses the generic kernel.

 > What if the generic kernel gets evicted?

It gets recreated upon next convolution primitive creation. 

>  How do we ensure deterministic behavior?

Define deterministic :-). In any case, as soon as a cache is used, the user is at the mercy of cache hit/miss. When using the hint, kernel compilation can be triggered at any time depending on both availability of kernel in the cache, and if the kernel for that specialization was already jitted or not.

I still think the async creation inside oneDNN poses a few problems (e.g. availability of threading runtime at creation, it breaks current model if users want to reuse primitive objects and cleanly separate creation overheads from execution).

Another option though would be to implement this strategy on FWK side. 
The generic primitive can be created using `DNNL_RUNTIME_DIM_VAL`, which would return unimplemented if we don't have it. They can also decide to use persistent cache APIs to make generic primitive creation faster.
Upon op execution, they can create specialized primitive asynchronously (respecting the FWKs threading runtime), and do something like this:

```
    auto specialized_future = std::async(std::launch::async, [] {
            return create_specialized_primitive(args);
        };
    auto generic_future= std::async(std::launch::async, [] {
            return create_generic_primitive(args);
        };

    auto primitive_to_run = wait_first_successful_complete(generic_future, specialize_future);
    execute_primitive(primitive_to_run.get(), args);
```

Reply:
> The specialized kernel gets evicted only only if all primitive objects using it are destroyed. 

But this is not how the kernel cache works. The cache holds kernels even if all the primitives that use them are destroyed. That's the point of the cache.

> Define deterministic :-).

For example, we define semantics of the runtime dimensions feature as follows. Pay the price of primitive creation once and then use the primitive for multiple shapes. With the primitive cache the users will have to pay the price of creation only once throughout workload.

When it comes to the asynchronous kernels creation it doesn't seem that we can guarantee anything because the kernels can get evicted at any time and therefore we will have to create them again or use the generic kernels which would lead to unexcepted performance effects.

Reply:
> But this is not how the kernel cache works. The cache holds kernels even if all the primitives that use them are destroyed. That's the point of the cache.

I meant it the otherway around, a kernel is not destroyed as long as there is a primitive object referencing it.

> With the primitive cache the users will have to pay the price of creation only once throughout the workload. This is guaranteed.

Is it? There is no such guarantee if the number of different primitives is higher than cache capacity. The behavior will both depend on distribution of primitive reuse, and eviction policy. 
The only difference between `generic + specialized` scheme vs `only generic` or `only specialized` schemes is the number of primitives/kernel to hold in cache.

Reply:
> Do you mean the existing kernel cache won't help with that?

Yes and no. It depends on how it gets implemented. If generic kernels are generated AOT, then indeed kernel cache won't help. If generic kernels are jitted, kernel cache is required to keep them alive.

> But what if we keep creating primitives for unique shapes? Does it mean that we will keep creating specialized kernels but using the generic ones?

Yes. In some sense, the approach can be summed up as opportunistic: by default we use generic kernels and opportunistically jit the specialized kernel. If we are lucky, it gets reused, if not it goes to waste. It boils down to a question of how much CPU usage matter for workloads running on GPU.

> Won't it just keep polluting the kernel cache causing the generic kernels to be evicted?

Yes it will pollute cache, but for each primitive created, we should bump the generic kernel as least recently used. This means there is a very low chance of evicting the generic kernel (unless cache capacity is smaller than number of generic kernels).

Reply:
I really like the idea of generic + specialized versions as this is a standard method for use in JIT compiled architectures. My initial reaction is that this sounds better than having an attribute for requesting reusable kernels. To some extent, the proposed attribute is intended to call reusable kernels before they are "feature/performance" complete, which is inherently problematic, while the generic + specialized solves the base issue. My main concern with this method is the work required to support a separate generic implementation, it does not sound trivial and it may be better to prioritize other implementations. 

In regards to some of the implementation related discussions:

> If generic kernels are generated AOT, then indeed kernel cache won't help.

This isn't quite true. It can still cost a (relatively) significant amount of time to transform a precompiled binary into a runtime object. Another point to consider is that, currently, accelerators performance is scaling much faster than general purpose CPU performance. As such, the CPU time required to dispatch work to accelerators is expected to become more critical.  Our current amortized target for calling a primitives is on the order of tens of microseconds, while creating a runtime object from binary alone currently take a few milliseconds (I did not investigate if it could be optimized though). As such, we already need to reuse runtime objects an average of 100s of times to hit that target.

In regards to storing generic kernels in the kernel cache, I think this is a bad idea.  Instead, my opinion is that we keep them in a separate fixed-size registry. As these are the generic implementations, there should not be very many actual kernels, and caching them just creates edges cases we can avoid at the cost of a small amount of (potentially) unused memory. I also think generic kernels need to be AOT compiled.  The idea around launching two expensive kernel creations on first primitive creation seems problematic.

Reply:
>  Another point to consider is that, currently, accelerators performance is scaling much faster than general purpose CPU performance. As such, the CPU time required to dispatch work to accelerators is expected to become more critical. 

There is an activity to address the problem via introducing SYCL Graph ([specification](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc)). We are collaborating with the team that is working on enabling SYCL Graph for IPEX.

Ultimately, we will have something like [CUDA Graph](https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/).